### PR TITLE
expression: change test for string_test.

### PR DESCRIPTION
**Pull request by hanfei1991**
add test for new interface step by step.
@coocood @zimulala @shenli @tiancaiamao  PTAL

### CODE REVIEW

File: expression/builtin.go [hunk 1]
```diff
@@ -18,28 +18,28 @@
...
 
 import (
-	"strings"
-
 	"github.com/juju/errors"
 	"github.com/pingcap/tidb/ast"
 	"github.com/pingcap/tidb/context"
 	"github.com/pingcap/tidb/parser/opcode"
 	"github.com/pingcap/tidb/util/types"
+	"strings"
```
Comment (tiancaiamao) [inline expression/builtin.go]: It's unnecessary to change this.

File: expression/builtin_string_test.go [hunk 2]
```diff
@@ -319,17 +319,19 @@ func (s *testEvaluatorSuite) TestSubstring(c *C) {
...
-		args := types.MakeDatums(v.str, v.pos)
+		fc := funcs[ast.Substring]
+		datums := types.MakeDatums(v.str, v.pos)
 		if v.slen != -1 {
-			args = append(args, types.NewDatum(v.slen))
+			datums = append(datums, types.NewDatum(v.slen))
 		}
-		r, err := f.F(args, s.ctx)
+		args := datumsToConstants(datums)
+		f, _ := fc.getFunction(args, s.ctx)
```
Comment (tiancaiamao) [inline expression/builtin_string_test.go]: Check error.

File: expression/builtin_time.go [hunk 3]
```diff
@@ -1445,6 +1445,26 @@ func parseDayInterval(sc *variable.StatementContext, value types.Datum) (int64,
 	return value.ToInt64(sc)
 }
 
+type timestampDiffFunctionClass struct {
+	baseFunctionClass
+}
+
+func (c *timestampDiffFunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
+	return &builtinTimestampDiffSig{newBaseBuiltinFunc(args, ctx)}, errors.Trace(c.verifyArgs(args))
```
Comment (tiancaiamao) [inline expression/builtin_time.go]: What does `deterministic` mean? Why it's true here but false in other places?

File: expression/builtin.go [hunk 4]
```diff
@@ -18,28 +18,28 @@
...
 
 func newBaseBuiltinFunc(args []Expression, ctx context.Context) baseBuiltinFunc {
 	return baseBuiltinFunc{
-		args:      args,
-		argValues: make([]types.Datum, len(args)),
-		ctx:       ctx,
+		args:          args,
+		argValues:     make([]types.Datum, len(args)),
+		ctx:           ctx,
+		deterministic: true,
```
Comment (tiancaiamao) [inline expression/builtin.go]: Would it be convenient to use false as the default value?

File: expression/builtin_time.go (hunk 3 above)
Comment (hanfei1991) [inline expression/builtin_time.go]: see the comment of `IsDeterministic`

**Comment by hanfei1991** [comment]
@tiancaiamao PTAL

**Comment by tiancaiamao** [comment]
LGTM

**Comment by coocood** [comment]
LGTM

**Pull request merged**